lib/Dialect/SPIRV/SPIRVDialect.cpp

```diff
@@ -53,6 +53,18 @@ SPIRVDialect::SPIRVDialect(MLIRContext *context)
 // Type Parsing
 //===----------------------------------------------------------------------===//
 
+// Forward declarations.
+template <typename ValTy>
+Optional<ValTy> parseAndVerify(SPIRVDialect const &dialect, Location loc,
+                               StringRef spec);
+template <>
+Optional<Type> parseAndVerify<Type>(SPIRVDialect const &dialect, Location loc,
```

Comment:
I don't think we need to forward declare the template specialization.

Reply:
Thanks, I just didn't want to move the functions body, since I've used it in array parser, early then definitions. So, it it's ok to move them.

Reply:
I didn't mean move them. You can forward declare the template function. I dont think you need to forward declare the specialization. So keep 
```
template <typename ValTy>
Optional<ValTy> parseAndVerify(SPIRVDialect const &dialect, Location loc,
                               StringRef spec);
```

but remove the 

```
template <>
Optional<Type> parseAndVerify<Type>(SPIRVDialect const &dialect, Location loc, StringRef spec)
```
and 

```
template <>
Optional<uint64_t> parseAndVerify<uint64_t>(SPIRVDialect const &dialect, Location loc, StringRef spec);
```

I think that should work. If I am mistaken, I am sorry about that then.

Reply:
Oh, sorry didn't get it right at first. Seems like in this case we will get an error like specilization of "function", after instantiation.

Reply:
Oh Ok. Then sorry about that. This looks fine then.